core/src/main/java/org/elasticsearch/cluster/routing/allocation/AllocationService.java

```diff
@@ -172,6 +184,44 @@ public Result applyFailedShards(ClusterState clusterState, List<FailedRerouteAll
         return buildResultAndLogHealthChange(allocation, "shards failed [" + failedShardsAsString + "] ...");
     }
 
+    private ClusterState removeStaleAllocationIds(ClusterState clusterState, List<FailedRerouteAllocation.StaleShard> staleShards) {
+        MetaData oldMetaData = clusterState.metaData();
+        MetaData.Builder metaDataBuilder = null;
+        for (Map.Entry<Index, List<FailedRerouteAllocation.StaleShard>> entry : staleShards.stream().collect(
+            Collectors.groupingBy(fs -> fs.shardId.getIndex())).entrySet()) {
+            final IndexMetaData oldIndexMetaData = oldMetaData.getIndexSafe(entry.getKey());
```

Comment:
We check the existence of the index in `ShardStateAction`, which ensures that every `FailedRerouteAllocation.StaleShard` entry references an index that exists (it's part of the contract of `applyFailedShards`), see https://github.com/ywelsch/elasticsearch/blob/ff17552b8d561232d7bfd51ea3ca8730d195fbe1/core/src/main/java/org/elasticsearch/cluster/action/shard/ShardStateAction.java#L260.

I can document this property on the `removeStaleAllocationIds` method as well.